packages/instrumentation-neo4j/src/neo4j.ts

```diff
@@ -0,0 +1,135 @@
+    InstrumentationNodeModuleFile,
+    InstrumentationNodeModuleDefinition,
+    safeExecuteInTheMiddle,
+} from '@opentelemetry/instrumentation';
+import { Neo4jInstrumentationConfig } from './types';
+import { getAttributesFromNeo4jSession } from './utils';
+
+type Neo4J = typeof neo4j;
+
+export class Neo4jInstrumentation extends InstrumentationBase<Neo4J> {
+    static readonly component = 'neo4j-driver';
+    protected _config!: Neo4jInstrumentationConfig;
+
+    constructor(config: Neo4jInstrumentationConfig = {}) {
+        super('opentelemetry-instrumentation-neo4j', VERSION, Object.assign({}, config));
+    }
+
+    setConfig(config: Neo4jInstrumentationConfig = {}) {
+        this._config = config;
+    }
+
+    protected init(): InstrumentationModuleDefinition<Neo4J> {
+        const apiModuleFiles = ['session', 'transaction'].map(
+            (file) =>
+                new InstrumentationNodeModuleFile<neo4j.Session>(
+                    `neo4j-driver/lib/${file}.js`,
+                    ['*'],
+                    this.patchSessionOrTransaction.bind(this),
+                    this.unpatchSessionOrTransaction.bind(this)
+                )
+        );
+
+        const module = new InstrumentationNodeModuleDefinition<Neo4J>(
+            Neo4jInstrumentation.component,
+            ['*'],
+            undefined,
+            undefined,
+            apiModuleFiles
+        );
+
+        return module;
+    }
+
+    private patchSessionOrTransaction(
+        fileExport: { default: () => neo4j.Session | neo4j.Transaction },
+        moduleVersion: string
+    ) {
+        const self = this;
+        this._wrap(fileExport.default.prototype, 'run', (originalRun: neo4j.Session['run']) => {
+            return function (query: string) {
+                if (self._config?.ignoreOrphanedSpans && !getSpan(context.active())) {
+                    return originalRun.apply(this, arguments);
+                }
+
+                const connectionAttributes = getAttributesFromNeo4jSession(this);
+                const operation = query.split(/\s+/)[0];
+                const span = self.tracer.startSpan(`Neo4j ${operation}`, {
+                    attributes: {
+                        ...connectionAttributes,
+                        [DatabaseAttribute.DB_SYSTEM]: 'neo4j',
+                        [DatabaseAttribute.DB_OPERATION]: operation,
+                        [DatabaseAttribute.DB_STATEMENT]: query,
+                    },
+                });
+                if (self._config.moduleVersionAttributeName) {
+                    span.setAttribute(self._config.moduleVersionAttributeName, moduleVersion);
+                }
+
+                const response: neo4j.Result = originalRun.apply(this, arguments);
+
+                const originalSubscribe = response.subscribe;
+                response.subscribe = function (observer) {
+                    const records = [];
+
+                    if (observer.onKeys) {
```

Comment:
is it possible for user to set `onKeys`, and then have the other callbacks called?
If so I think it's safer to always patch all callbacks

Reply:
according to docs they don't go together
![image](https://user-images.githubusercontent.com/16322616/109770435-3f618080-7c04-11eb-9751-fc6eb9bd4aef.png)

Reply:
the following call will patch only the `onKeys`, even though the `onNext` -> `onCompleted` flow might be the actual one being called:
```js
session
  .run('MERGE (alice:Person {name : $nameParam}) RETURN alice.name AS name', {
    nameParam: 'Alice'
  })
  .subscribe({
    onKeys: keys => {
      console.log(keys)
    },
    onNext: record => {
      console.log(record.get('name'))
    },
    onCompleted: () => {
      session.close() // returns a Promise
    },
    onError: error => {
      console.log(error)
    }
  })
```